damnit/backend/db.py

```diff
@@ -41,6 +42,26 @@ class ReducedData:
     max_diff: float = None
 
 
+class BlobTypes(Enum):
+    png = 'png'
+    numpy = 'numpy'
+    pickle = 'pickle'  # We try to avoid pickle, but have used it in the past
+    unknown = 'unknown'
+
+    @classmethod
+    def identify(cls, blob: bytes):
+        if blob.startswith(b'\x89PNG\r\n\x1a\x0a'):
+            return cls.png
+        elif blob.startswith(b'\x93NUMPY'):
+            return cls.numpy
+        elif blob.startswith(b'\x80'):
+            # Since pickle v2 (Python 2.3), all pickles start with a protocol
+            # version opcode (0x80).
+            return cls.pickle
```

Comment:
I'm looking at this roughly as: anything that's not just strings and numbers can be stored in the database as blobs. We can distinguish types within that by tagging them with a unique prefix to identify them, and many formats helpfully include such a prefix already. It's not a hack, IMO, because formats include these 'signatures' precisely so you can recognise the data, and they're part of the specs ([e.g. PNG](http://libpng.org/pub/png/spec/iso/index-object.html#5PNG-file-signature)).

Pickle comes with security concerns - unpickling is effectively execution. But also, pickled data saved with one version of the code may not work with a later version. You can make it work if you're careful enough, but it's easy to slip up, and the usual advice is to use pickle only for sending data over the wire (between trusted components!), and not for storing data.

msgpack: if we need it :shrug: . What do you envisage storing in it? Do you mean as another option in addition to PNG and `numpy.save()`, or replace all blobs with msgpack?

Reply:
Hmm, ok. I don't particularly like relying on file headers just in case we want to use some type that has a conflicting header with another type. But I don't think we're going to store that many types different types anyway so it's probably fine.